odl/trafos/wavelet.py

```diff
@@ -21,9 +21,836 @@
+            ``pywt.MODES.modes``
+        nscales : `int`
+            Number of scales in the coefficient list
+
+    Returns
+    -------
+        x : `numpy.ndarray`.
+        A wavalet reconstruction.
+
+    See also
+    --------
+    pywt : http://www.pybytes.com/pywavelets/
+    """
+    aaa = coeff_list[0]
+    k = 1
+    (aad, ada, add, daa, dad, dda, ddd) = coeff_list[k]
+    coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
+                  'daa': daa, 'dad': dad, 'dda': dda, 'ddd': ddd}
+    for k in range(2, nscales+1):
+        aaa = pywt.idwtn(coeff_dict, wbasis, mode)
+        (aad, ada, add, daa, dad, dda, ddd) = coeff_list[k]
+        coeff_dict = {'aaa': aaa, 'aad': aad, 'ada': ada, 'add': add,
+                      'daa': daa, 'dad': dad, 'dda': dda, 'ddd': ddd}
+
+    x = pywt.idwtn(coeff_dict, wbasis, mode)
+    return x
+
+
+class DiscreteWaveletTrafo(Operator):
+
+    """Discrete wavelet trafo between discrete L2 spaces."""
+
+    def __init__(self, dom, nscales, wbasis, mode):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        dom : :class:`~odl.discr.lp_discr.DiscreteLp`
+            Domain of the wavelet transform (the "image domain").
+            The exponent :math:`p` of the discrete :math:`L^p`
+            space must be equal to 2.0.
+        nscales : `int`
+            Number of scales in the coefficient list
+        wbasis :  ``_pywt.Wavelet``
+            Describes properties of a selected wavelet basis
+        mode : `str`
+            Signal extention mode. For possible extensions see
+            ``pywt.MODES.modes``
+
+        See also
+        --------
+        pywt : http://www.pybytes.com/pywavelets/
+        """
+        self.nscales = int(nscales)
+        self.wbasis = wbasis
+        self.mode = str(mode).lower()
+
+        max_level = pywt.dwt_max_level(dom.grid.shape[0],
+                                       filter_len=self.wbasis.dec_len)
+        # TODO: maybe the error message could tell how to calculate the
+        # max number of levels
+        if self.nscales >= max_level:
+            raise ValueError('Cannot use more than {} scaling levels, '
+                             'got {}.'.format(max_level, self.nscales))
+
+        self.size_list = list_of_coeff_sizes(dom.grid.shape, self.nscales,
+                                             self.wbasis, self.mode)
+
+        ran_size = np.prod(self.size_list[0])
+
+        if len(dom.grid.shape) == 2:
+            ran_size += sum(3 * np.prod(shape) for shape in
+                            self.size_list[1:-1])
+        if len(dom.grid.shape) == 3:
+            ran_size += sum(7 * np.prod(shape) for shape in
+                            self.size_list[1:-1])
+
+        # TODO: Maybe allow other ranges like Besov spaces (yet to be crated)
+        ran = dom.dspace_type(ran_size, dtype=dom.dtype)
+        super().__init__(dom, ran)
+
+        if not isinstance(dom, DiscreteLp):
```

Comment:
Move test to top of function, you assume `dom.grid.shape` exists before, this would give a `TAttributeError` before we even got here.